transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -400,74 +383,125 @@ private int findCtxIdx(Predicate<AbstractChannelHandlerContext> predicate) {
     @Override
     public final ChannelPipeline remove(ChannelHandler handler) {
         final AbstractChannelHandlerContext ctx;
+        EventExecutor executor = executor();
+        boolean inEventLoop = executor.inEventLoop();
         synchronized (handlers) {
-            int idx = findCtxIdx((context) -> context.handler() == handler);
+            int idx = findCtxIdx(context -> context.handler() == handler);
             if (idx == -1) {
                 throw new NoSuchElementException();
             }
             ctx = handlers.remove(idx);
+            assert ctx != null;
+
+            if (!inEventLoop) {
```

Comment:
Technically I don't think this issue applies to the remove cases - the same kind of interleaving won't matter since the end result will be the same. So the execute could remain outside the sync block, but maybe it's better like this anyhow for consistency

Reply:
Actually yes I guess better to keep like this so the rollback case is handled properly